cloudevents/working-drafts/minor-versions.md

```diff
@@ -0,0 +1,337 @@
+> in the serialization. Note: for 'release candidate' releases a
+> suffix might be used for testing purposes.
+
+The GOVERNANCE.md doc states:
+
+> Since changing the CloudEvents specversion string could have a
+> significant impact on implementations, all non-breaking changes will
+> be made as "patch" version updates - this allows for the value "on
+> the wire" to remain unchanged. If a breaking change is introduced
+> the normal semver rules will apply and the "major" version number
+> will change. The net effect of this is that the "minor" version
+> number will always be zero and the specversion string will always be
+> of the form X.0.
+
+I'd view the governance doc as rather less binding than the spec -
+it's "current practice" rather than a promise.
+
+Any SDK (and potentially other consuming code) that *expects* the
+current documented will fail to parse a v1.1 (etc) event - and
+that's probably okay. For maximum compatibility, publish v1.0
+events for a while. But we don't even need a v1.1 right now, so
+there's enough time for SDKs to start implementing the below
+compatibility code *now*, and so hopefully by the time we actually
+want a v1.1, the majority of deployed software will be able to
+handle it.
+
+## Out of scope (major)
+
+There are probably many changes which could count as major, but I'd
+like to call out the following explicitly:
+
+- Any change to required attributes:
+  - Adding
+  - Removing
+  - Changing type
+  - Changing validation criteria
+- Changes to the type system:
+  - Adding types
+  - Removing types
+- Changes to attribute naming rules
+- Breaking changes to CloudEvent formats
+
+## Intended minor change
+
+Fundamentally, the purpose of this proposal is to allow the
+introduction of new *optional* attributes, which are
+expected to usually have been documented extension attributes in the
+past - in other words, a graduation process from "suggested as useful"
+to "officially specified". However, there's no requirement for an
+attribute to be a documented extension attribute before being added;
+that's just an informal part of the usual process.
+
+SemVer minor versions are intended for new features that don't break
+existing code. How SemVer applies to specifications is less clear-cut
+than how it applies to actual code. This document attempts to imagine
+the impact of a change on various stakeholders.
+
+For the purpose of this document, we'll consider a single new optional
+attribute into version 1.1:
+
+- Name: `xyz`
+- Type: integer
+- Constraints: value in the range 1-100 inclusive
+
+Obviously this is entirely artificial, but it should be enough to help
+find corner cases and problems.
+
+## Logical compatibility
+
+Every valid v1.1 event can be "rewritten" as a valid v1.0 event, but the reverse is not true.
+A v1.0 event containing an `xyz` extension attribute with value "test" is valid,
+but cannot be expressed as a valid v1.1 event. As there are no restrictions on
```

Comment:
True but we do (sort of) warn people about picking a name that might be used in the future.
> New extension definitions SHOULD use a name that is descriptive enough to reduce the chances of name collisions with other extensions. In particular, extension authors SHOULD check the [documented extensions](https://github.com/cloudevents/spec/blob/main/cloudevents/documented-extensions.md) document for the set of known extensions - not just for possible name conflicts but for extensions that might be of interest.

Granted, this talks about other extensions but I believe it was meant to also cover possible future CE defined attributes. Perhaps we need to call that out.

My point is that adding extensions is always a risky business if you don't choose a name that's unique or has some scoping mechanism/namespace that others won't use.  Not saying we shouldn't do what you're suggesting, just saying that this is the nature of extensions and a version of this problem will always exist no matter what we do.

Reply:
Note that even HTTP itself has a variant of this problem since new RFC defined headers can be created at any time that overlap with an extension header. And people just deal with it. At some point this might come down to "is it worth the effort".

Reply:
They dropped the use of X-* headers, I believe, for many of the same reasons we didn't create an "extension" bag of attributes. It's a false security blanket and eventually causes pain as things move from being extensions to core.